crates/core/c8y_smartrest/src/operations.rs

```diff
@@ -0,0 +1,215 @@
+        .map(|entry| entry.map(|e| e.path()))
+        .collect::<Result<Vec<PathBuf>, _>>()?
+        .into_iter()
+        .filter(|path| path.is_file())
+        .collect::<Vec<PathBuf>>();
+
+    for path in dir_entries {
+        let mut details = match fs::read(&path) {
+            Ok(bytes) => toml::from_slice::<Operation>(bytes.as_slice())
+                .map_err(|e| OperationsError::TomlError(path.to_path_buf(), e))?,
+
+            Err(err) => return Err(OperationsError::FromIo(err)),
+        };
+
+        details.name = path
+            .file_name()
+            .and_then(|filename| filename.to_str())
+            .ok_or_else(|| OperationsError::InvalidOperationName(path.to_owned()))?
+            .to_owned();
+
+        operations.add(details);
+    }
+    Ok(operations)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::io::Write;
+
+    use super::*;
+    use test_case::test_case;
+
+    // Structs for state change with the builder pattern
+    // Structs for Operations
+    struct Ops(Vec<PathBuf>);
+    struct NoOps;
+
+    struct TestOperationsBuilder<O> {
+        temp_dir: tempfile::TempDir,
+        operations: O,
+    }
+
+    impl TestOperationsBuilder<NoOps> {
+        fn new() -> Self {
+            Self {
+                temp_dir: tempfile::tempdir().unwrap(),
+                operations: NoOps,
+            }
+        }
+    }
+
+    impl TestOperationsBuilder<NoOps> {
+        fn with_operations(self, operations_count: usize) -> TestOperationsBuilder<Ops> {
+            let Self { temp_dir, .. } = self;
+
+            let mut operations = Vec::new();
+            for i in 0..operations_count {
+                let file_path = temp_dir.path().join(format!("operation{}", i));
+                let mut file = fs::File::create(&file_path).unwrap();
+                file.write_all(
+                    br#"[exec]
+                        command = "echo"
+                        on_message = "511""#,
+                )
+                .unwrap();
+                operations.push(file_path);
+            }
+
+            TestOperationsBuilder {
+                operations: Ops(operations),
+                temp_dir,
+            }
+        }
+    }
+
+    impl TestOperationsBuilder<Ops> {
+        fn build(self) -> TestOperations {
+            let Self {
+                temp_dir,
+                operations,
+            } = self;
+
+            TestOperations {
+                temp_dir,
+                operations: operations.0,
+            }
+        }
+    }
+
+    struct TestOperations {
+        temp_dir: tempfile::TempDir,
+        operations: Vec<PathBuf>,
+    }
+
+    impl TestOperations {
+        fn builder() -> TestOperationsBuilder<NoOps> {
+            TestOperationsBuilder::new()
+        }
+
+        fn temp_dir(&self) -> &tempfile::TempDir {
+            &self.temp_dir
+        }
+    }
+
+    #[test_case(0)]
+    #[test_case(1)]
+    #[test_case(5)]
+    fn get_operations_all(ops_count: usize) {
```

Comment:
I'm not convinced that this test brings a lot of value.

Reply:
I think this was a previous commit, I will try to improve git diffs in future PRs